### PR TITLE
Add support for OpenShift

### PR DESCRIPTION
**Pull request by amrut-asm**
OpenShift requires securityContext to be set for pods in order to decide an appropriate SCC to apply to them.

This PR:

1. Adds SecurityContext for the model pods created by kubeAI
2. Adds SecurityContext for the model containers created by kubeAI
3. Sets securityContext for openwebui container
4. Creates securityContextConstraints (SCC) for openwebui and the model pods

Chart version fields are untouched... not sure what to do with them

### CODE REVIEW

**Comment by nstogner** [comment]
Thanks for the PR! We will take a look.

File: api/v1/model_types.go
Comment (nstogner) [inline api/v1/model_types.go:50]: We would like to keep the Model spec as simple as possible. I think it would be better to have these fields defined at the system-level (in the ConfigMap for KubeAI instead of the Model spec). Perhaps these would belong to a field inside of the `internal/config` pkg's `System` struct:

```go
type System struct {
    // ...
    ModelServerPods ModelServerPods `json:"modelServerPods"`
}
```

Let me know if there is any reason why defining the same values across all model Pods would be an issue.

**Review by nstogner — CHANGES_REQUESTED** [review]
Thanks for taking the time to submit the PR! Added a comment about where we would prefer to specify these fields.

File: api/v1/model_types.go
Comment (samos123) [reply api/v1/model_types.go:50]: What's your thought on this instead?
```golang
type System struct {
         // The service account to use for all model pods
	ModelServiceAccountName string `json:"serviceAccountName,omitempty"`

	// Security Context for the model pods
	ModelPodSecurityContext *corev1.PodSecurityContext `json:"podSecurityContext,omitempty"`

	// Security Context for the model pod containers
	ModelContainerSecurityContext *corev1.SecurityContext `json:"securityContext,omitempty"`
}
```

**Comment by samos123** [comment]
Thank you for this PR! We think it would be very valuable to have OpenShift support and will try to merge this once changes are addressed.

The chart versions do not need to be bumped. We currently do manual releases with more testing. This allows us to merge changes into main without being afraid of breaking users.

It would also be very helpful to have an OpenShift specific installation guide once this PR gets merged or as part of this PR.

Feel free to also join the discord and we can discuss relatime if needed.

File: api/v1/model_types.go
Comment (samos123) [reply api/v1/model_types.go:50]: I see and agree. I didn't  get your initial comment.

**Comment by samos123** [comment]
Saw you closed the PR, but we're more than happy to take this PR for which already did most of the work and make the requested changes directly. You would still be author. Please let me know if you would prefer that. Apologies for not being able to take the PR as-is directly, because I do really think this would be valuable.

**Comment by amrut-asm** [comment]
I discarded the commits and synced my branch with main and it somehow closed the PR 😔. I have made the changes already according to the new format, testing it currently. Will re-open this if possible otherwise I'll create a new PR 🙂

**Comment by amrut-asm** [comment]
On the OpenShift specific installation guide - Nothing out of the ordinary... following the regular installation process will work fine

**Comment by amrut-asm** [comment]
> Thank you for this PR! We think it would be very valuable to have OpenShift support and will try to merge this once changes are addressed.
> 
> The chart versions do not need to be bumped. We currently do manual releases with more testing. This allows us to merge changes into main without being afraid of breaking users.
> 
> It would also be very helpful to have an OpenShift specific installation guide once this PR gets merged or as part of this PR.
> 
> Feel free to also join the discord and we can discuss relatime if needed.


On the OpenShift specific installation guide - Nothing out of the ordinary... following the regular installation process will work fine

File: api/v1/model_types.go
Comment (amrut-asm) [reply api/v1/model_types.go]: I've made the necessary modifications.

File: charts/kubeai/values.yaml
Comment (samos123) [inline charts/kubeai/values.yaml:36]: Just curious, no change needed. Have you seen this cause issues for pods?

I don't expect the LLM engines to fail but there is always a risk of issues when you enable more security settings by default. 

I do like having secure defaults that work! So I'm in favor of keeping this as-is.

File: charts/kubeai/templates/configmap.yaml
Comment (samos123) [inline charts/kubeai/templates/configmap.yaml:25]: What should be the behavior if someone sets `.Values.modelServiceAccount.name` and creates this K8s SA themselves?

I think in that case the serviceAccountName should be set.

I think this if should be removed and then we need to ensure that `.Values.modelServiceAccount.name` is `default` and `create=false` by default.

Comment (samos123) [reply charts/kubeai/templates/configmap.yaml:25]: Note we do this as a follow up because create=true by default. So this is more of an edge case.

File: charts/kubeai/templates/securityContextConstraints.yaml
Comment (samos123) [inline charts/kubeai/templates/securityContextConstraints.yaml:2]: same comment in regards to setting custom service account and create=false

**Review by samos123 — APPROVED** [review]
I left some comments but LGTM overall to merge and we can tackle these as follow ups or you could.

Thank you for adding OpenShift support and improving out of the box security! Will wait for @nstogner to finish his review too.

File: charts/kubeai/charts/openwebui/templates/securityContextConstraints.yaml
Comment (nstogner) [inline charts/kubeai/charts/openwebui/templates/securityContextConstraints.yaml:6]: Let's template this with fullname like other resources.

File: charts/kubeai/templates/securityContextConstraints.yaml
Comment (nstogner) [inline charts/kubeai/templates/securityContextConstraints.yaml:6]: Let's template this with fullname like other resources.

**Review by nstogner — CHANGES_REQUESTED** [review]
Looking good - a minor things.

**Comment by amrut-asm** [comment]
> Looking good - a minor things.

I'll address the issues and fix them soon 🙂

File: charts/kubeai/templates/configmap.yaml
Comment (amrut-asm) [reply charts/kubeai/templates/configmap.yaml:25]: > What should the behavior be if someone sets `.Values.modelServiceAccount.name` with `create=false` and creates this K8s SA themselves without using helm?
> 
> I think in that case the serviceAccountName should be set.
> 
> I think this if statement could be removed?

Yes, I agree. This statement is not needed.

File: charts/kubeai/templates/securityContextConstraints.yaml
Comment (amrut-asm) [reply charts/kubeai/templates/securityContextConstraints.yaml:2]: Agree

File: charts/kubeai/templates/_helpers.tpl
Comment (amrut-asm) [reply charts/kubeai/templates/_helpers.tpl:71]: Sorry, I don't understand. It is defaulting to "default" when 
1. modelServiceAccount.create is false
2. modelServiceACcount.name is empty ""

Is it not expected?

File: charts/kubeai/templates/securityContextConstraints.yaml
Comment (amrut-asm) [reply charts/kubeai/templates/securityContextConstraints.yaml:5]: Done

File: charts/kubeai/values.yaml
Comment (amrut-asm) [reply charts/kubeai/values.yaml:36]: > Just curious, no change needed. Have you seen this cause issues for pods in general (not on KubeAI)?
> 
> I don't expect the LLM engines to fail but there is always a risk of issues when you enable more security settings by default.
> 
> I do like having secure defaults that work! So I'm in favor of keeping this as-is.

On OpenShift, from what I've observed - a container must set it's securityContext according to the privileges it requires so as to be assigned an appropriate SCC that matches it's securityContext settings. Failing to do so (for example, not setting securityContext at all or not matching the securityContext for the SCC which we are *hoping* to be assigned) will result in the default SCC to be assigned (which happens to be restricted-v2).

So, yeah this causing issues for pods is only true on OpenShift and not for others.

Comment (samos123) [reply charts/kubeai/values.yaml:36]: Thank a lot for explaining. I get it now.

File: charts/kubeai/charts/openwebui/templates/securityContextConstraints.yaml
Comment (amrut-asm) [reply charts/kubeai/charts/openwebui/templates/securityContextConstraints.yaml]: Done

File: charts/kubeai/templates/securityContextConstraints.yaml
Comment (amrut-asm) [reply charts/kubeai/templates/securityContextConstraints.yaml]: Done

File: charts/kubeai/templates/_helpers.tpl
Comment (samos123) [reply charts/kubeai/templates/_helpers.tpl:71]: It's correct as-is: https://helm.sh/docs/chart_template_guide/function_list/#default

no need to change it.

**Review by samos123 — APPROVED** [review]

**Pull request merged**
